### PR TITLE
Connect the map to Calendar, with date picker

### PR DESCRIPTION
**Pull request by ahlschlager**

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/SortedMarkersServlet.java [hunk 1]
```diff
@@ -49,41 +50,34 @@ private String getParameter(HttpServletRequest request, String name, String defa
...
+        for (Event calEvent : calEvents.getItems()) {
+            EventMarker mapEvent = new EventMarker(
+                calEvent.getSummary(),
+                calEvent.getDescription(),
+                calEvent.getLocation(),
+                calEvent.getStart().getDateTime(),
+                "Black Lives Matter");
+            mapEvents.add(mapEvent);
+        }
+        SortedMarkers sorted = new SortedMarkers(mapEvents);
```
Comment (vlop11) [inline src/main/java/com/google/sps/servlets/SortedMarkersServlet.java]: I would name this something a little more descriptive. maybe sortedEvents or sortedMarkers?

File: src/main/java/com/google/sps/utils/StringConstants.java [hunk 2]
```diff
@@ -31,6 +33,10 @@
   public static final String SUMMARY_SCORE = "summaryScore";
   public static final String TIMESTAMP_PROPERTY = "timestamp";
   public static final String VALUE = "value";
+  public static final String CHAT_PAGE_NAMESPACE = "templates.chat.chatPage";
+  public static final String KEYS_TXT_FILE = "keys.txt";
+  public static final String CHAT_ERROR_NAMESPACE = "templates.chat.error";
+  public static final String REFERER_HEADER = "Referer";
```
Comment (vlop11) [inline src/main/java/com/google/sps/utils/StringConstants.java]: look at Dylan's file for ordering, he put all the template names at the top and keys and referer should be in alphabetical order in the chat constants section

File: src/main/java/com/google/sps/servlets/SortedMarkersServlet.java [hunk 3]
```diff
@@ -49,41 +50,34 @@ private String getParameter(HttpServletRequest request, String name, String defa
...
-        events.add(event3);
-
-        SortedMarkers sorted = new SortedMarkers(events); 
-        return sorted.getSortedMarkers();
+    try {
+        ArrayList<EventMarker> mapEvents = new ArrayList<EventMarker>();
+        EventsServlet eventsServlet = new EventsServlet();
+        
+        // CalendarID and group Name are both hardcoded for now
+        // The events are from the calendar though
```
Comment (sporkert) [inline src/main/java/com/google/sps/servlets/SortedMarkersServlet.java]: You can take out this line

File: src/main/java/com/google/sps/servlets/SortedMarkersServlet.java [hunk 4]
```diff
@@ -49,41 +50,34 @@ private String getParameter(HttpServletRequest request, String name, String defa
...
-        events.add(event2);
-        events.add(event3);
-
-        SortedMarkers sorted = new SortedMarkers(events); 
-        return sorted.getSortedMarkers();
+    try {
+        ArrayList<EventMarker> mapEvents = new ArrayList<EventMarker>();
+        EventsServlet eventsServlet = new EventsServlet();
+        
+        // CalendarID and group Name are both hardcoded for now
```
Comment (sporkert) [inline src/main/java/com/google/sps/servlets/SortedMarkersServlet.java]: Can you rephrase this to something like TODO: Replace hard-coded calendarID and group name when ____

That way it's more clear to the reader what you're waiting for

File: src/main/java/com/google/sps/servlets/MapServlet.java [hunk 5]
```diff
@@ -11,44 +14,35 @@
...
-        
-        String out = tofu
-            .newRenderer("templates.mapPages.mapPage")
-            // TODO change to constants from Valeria's StringConstants.java for the strings once her PR is merged
-            .setData(ImmutableMap.of("groups", ImmutableList.of(groupMap, groupMap2), "key", apiKey))
-            .render();
-        response.getWriter().println(out);
+ 
+        final String mapPageHtml = getOutputString(MAP_SOY_FILE, MAP_TEMPLATE_NAMESPACE, 
+            ImmutableMap.of(GROUPS_KEY, ImmutableList.of(groupMap, groupMap2), "key", apiKey));
```
Comment (sporkert) [inline src/main/java/com/google/sps/servlets/MapServlet.java]: replace "key" with a constant

**Review by sporkert — CHANGES_REQUESTED** [review]
Can you check the formatting? Maybe something happened when you were moving files over.

There's a lot of lines that show just a space added.

File: src/main/java/com/google/sps/EventMarker.java [hunk 6]
```diff
@@ -17,36 +17,47 @@
     private String description;
     private String location; 
     private DateTime date; 
+    private String simpleDate;
     private String dateOutput; 
     private String groupName;
-    private static final DateTimeFormatter format = DateTimeFormatter.ofPattern("EEEE, MMMM dd, yyyy hh:mm a");
-
+ 
+    private static final DateTimeFormatter FORMAT = DateTimeFormatter.ofPattern("EEEE, MMMM dd, yyyy hh:mm a");
```
Comment (sporkert) [inline src/main/java/com/google/sps/EventMarker.java]: private static final variables should go at the top

File: src/main/java/com/google/sps/EventMarker.java [hunk 7]
```diff
@@ -17,36 +17,47 @@
     private String description;
```
Comment (sporkert) [inline src/main/java/com/google/sps/EventMarker.java:21]: I know this isn't part of your change, but while you're here could you alphabetize these?

File: src/main/java/com/google/sps/EventMarker.java [hunk 8]
```diff
@@ -17,36 +17,47 @@
...
     private void formatDate() {      
-        dateOutput = OffsetDateTime.parse(date.toStringRfc3339()).format(format); 
+        dateOutput = OffsetDateTime.parse(date.toStringRfc3339()).format(FORMAT); 
     }
-
+ 
+    /**
+     * Formats the Date to a simplified version: 2020-08-13
+     */
+    private void simpleDate() {
```
Comment (sporkert) [inline src/main/java/com/google/sps/EventMarker.java]: rename this to formatSimpleDate()?

File: src/main/java/com/google/sps/servlets/MapServlet.java [hunk 9]
```diff
@@ -1,54 +1,81 @@
...
+        // Stores the IDs for all the groups the user ahs joined
+        HashSet<Long> userGroups = new HashSet<Long>();
+
+        try {
+            Entity user = datastore.get(KeyFactory.createKey(USER_KIND, userEmail));
+            userGroups = (HashSet<Long>) user.getProperty(GROUPS_KEY);
+
+        } catch (EntityNotFoundException newUser) {
+            // if the user does not exist, create one and add to datastore
+            Entity user = new Entity(USER_KIND, userEmail);
```
Comment (sporkert) [inline src/main/java/com/google/sps/servlets/MapServlet.java]: Is there a way to detect if a user does not exist without waiting for an exception to be thrown?

Exceptions should really be used for actual exceptions that you can't really predict. If you can predict it (which it looks like you can here), use an if/else and handle the edge case.

**Review by sporkert — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/EventMarker.java (hunk 7 above)
Comment (sporkert) [reply src/main/java/com/google/sps/EventMarker.java:21]: Thanks!

File: src/main/java/com/google/sps/EventMarker.java [hunk 10]
```diff
@@ -8,45 +8,56 @@
...
         return name;
     }
-
+ 
     /**
      * Formats the Date ex. Monday, June 08, 2000 12:30PM
      */
     private void formatDate() {      
-        dateOutput = OffsetDateTime.parse(date.toStringRfc3339()).format(format); 
+        dateOutput = OffsetDateTime.parse(date.toStringRfc3339()).format(FORMAT);
```
Comment (sporkert) [inline src/main/java/com/google/sps/EventMarker.java:45]: Are there tests already for date formatting? If not, please add :)

File: src/main/java/com/google/sps/servlets/SortedMarkersServlet.java [hunk 11]
```diff
@@ -49,41 +50,33 @@ private String getParameter(HttpServletRequest request, String name, String defa
...
+                calEvent.getSummary(),
+                calEvent.getDescription(),
+                calEvent.getLocation(),
+                calEvent.getStart().getDateTime(),
+                "Black Lives Matter");
+            mapEvents.add(mapEvent);
+        }
+        SortedMarkers sortedEvents = new SortedMarkers(mapEvents); 
+        return sortedEvents.getSortedMarkers();
+
```
Comment (sporkert) [inline src/main/java/com/google/sps/servlets/SortedMarkersServlet.java]: nit: remove extra newline

File: src/main/java/com/google/sps/utils/ServletUtils.java [hunk 12]
```diff
@@ -55,12 +61,57 @@ public static void printBadRequestError(HttpServletResponse response, String err
...
+    // Stores the IDs for all the groups the user ahs joined
+    HashSet<Long> userGroups = new HashSet<Long>();
+
+    Query userQuery = new Query(USER_KIND).addFilter(Entity.KEY_RESERVED_PROPERTY, FilterOperator.EQUAL, userEmail);
+    PreparedQuery userPreparedQuery = datastore.prepare(userQuery);
+    Entity user = userPreparedQuery.asSingleEntity();
+
+    if (user != null) {
+      userGroups = (HashSet<Long>) user.getProperty(GROUPS_KEY);
+
```
Comment (sporkert) [inline src/main/java/com/google/sps/utils/ServletUtils.java]: remove newline

File: src/main/java/com/google/sps/utils/ServletUtils.java [hunk 13]
```diff
@@ -55,12 +61,57 @@ public static void printBadRequestError(HttpServletResponse response, String err
...
+            GROUP_CALENDARID_PROPERTY, (String) group.getProperty(GROUP_CALENDARID_PROPERTY),
+            GROUP_IMAGE_PROPERTY, (String) group.getProperty(GROUP_IMAGE_PROPERTY));
+        groupsListBuilder.add(groupMap);
+      } catch (EntityNotFoundException invalidGroup) {
+        continue;
+      }
+            
+    }
+    ImmutableList<ImmutableMap<String, String>> groupsList = groupsListBuilder.build();
+    return groupsList;
```
Comment (sporkert) [inline src/main/java/com/google/sps/utils/ServletUtils.java]: you can condense this into one line and just do `return groupsListBuilder.build();`

**Review by sporkert — CHANGES_REQUESTED** [review]

File: src/main/webapp/mapScript.js [hunk 14]
```diff
@@ -172,27 +182,63 @@ function addMarkerToMap(address, descriptionText, nameText, groupName, date, loc
...
+        if (dateInput != markers[i].date) {
+            markers[i].marker.setMap(null);
+        } else {
+            markers[i].marker.setMap(map);
+        }
+    }
+}
+ 
+function clearDate() {
+    document.getElementById("date").value = "";
```
Comment (sporkert) [inline src/main/webapp/mapScript.js:239]: use constant variable for `date`

File: src/main/java/com/google/sps/utils/ServletUtils.java [hunk 15]
```diff
@@ -55,12 +61,60 @@ public static void printBadRequestError(HttpServletResponse response, String err
...
+        Entity group = datastore.get(KeyFactory.createKey(GROUP_KIND, String.valueOf(groupId)));
+        ImmutableMap<String, String> groupMap = ImmutableMap.of(
+            GROUP_NAME_PROPERTY, (String) group.getProperty(GROUP_NAME_PROPERTY),
+            GROUP_ID_PROPERTY, group.getKey().getName(),
+            GROUP_DESCRIPTION_PROPERTY, (String) group.getProperty(GROUP_DESCRIPTION_PROPERTY),
+            GROUP_CALENDARID_PROPERTY, (String) group.getProperty(GROUP_CALENDARID_PROPERTY),
+            GROUP_IMAGE_PROPERTY, (String) group.getProperty(GROUP_IMAGE_PROPERTY));
+        groupsListBuilder.add(groupMap);
+      } catch (EntityNotFoundException invalidGroup) {
+        System.out.println("SYDNEY --- THE GROUP IS INVALID");
```
Comment (sporkert) [inline src/main/java/com/google/sps/utils/ServletUtils.java]: remove your name? :)

**Review by sporkert — APPROVED** [review]

File: src/main/webapp/mapScript.js (hunk 14 above)
Comment (ahlschlager) [reply src/main/webapp/mapScript.js:239]: Resolved offline

**Review by sporkert — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/MapServlet.java [hunk 16]
```diff
@@ -1,54 +1,48 @@
...
     /**
      * Adds user's groups as checkboxes to the map drawer
      * Allows users to select which groups they want displayed on the map
      */
     @Override
     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-
+        ServletUtils.enforceUserLogin(request, response);
+ 
         // Reads API Key and HTTP referer from file
```
Comment (vlop11) [inline src/main/java/com/google/sps/servlets/MapServlet.java]: you might want to change your comment here, since you're no longer reading the referer

**Review by vlop11 — APPROVED** [review]
Left a couple comments, but nothing major
